### ISSUE TITLE
Adding layers changes existing weights

### ISSUE THREAD

**Issue by m-toman** [tiny-dnn/tiny-dnn#616]
Hi,

I've noticed that adding layers after (manually) setting weights, modifies the weights of existing layers:

Adding a layer, then setting w[0]
```
dnnModel << tiny_dnn::fc<tiny_dnn::tan_h>(pbLayer.numinputs(), pbLayer.numoutputs());
auto layerWeights = layer.weights();
auto& w = *layerWeights[0];
w[0] =  -0.0511331186;
		
Debugger: 
   (w)[0]	-0.0511331186	float
```
After all weights for this layer are set, I'll check again:
```      
auto layerWeightsx = dnnModel[0]->weights();
auto& wx = *layerWeightsx[0];

Debugger:
   (wx)[0]	-0.0511331186	float
```

Adding a second layer and checking the weight of layer 0:
```
dnnModel << tiny_dnn::fc<tiny_dnn::tan_h>(pbLayer.numinputs(), pbLayer.numoutputs());
auto layerWeightsx = dnnModel[0]->weights();
auto& wx = *layerWeightsx[0];

Debugger:
   (w)[0]	0.0381551981	float
   (wx)[0]	0.0381551981	float
```

This does not occur when setting
```
layer.set_trainable(false);
```
after setting the weights and before adding the next layer.

Is this intended?

(Gitter links:
https://gitter.im/tiny-dnn/users?at=58c2fd97dd08b4b859b24cb5
and
https://gitter.im/tiny-dnn/users?at=58c3b0311465c46a56188983
)

**Comment by edgarriba** [comment]
Could you share the full snippet so that we can test it?

**Comment by m-toman** [comment]
This is integrated in a larger application that loads the weights from a protobuf file etc., but I'll see that I set up a small program asap.

**Comment by bhack** [comment]
Yes just a snippet to compile and reproduce it.

**Comment by m-toman** [comment]
Here we go:

```
#include <iostream>
#include "tiny_dnn/tiny_dnn.h"

void showcase(bool trainable) {
   tiny_dnn::network<tiny_dnn::sequential> net;

   net << tiny_dnn::fc<tiny_dnn::tan_h>(2, 2, false);
   auto& layer = *net[net.layer_size() - 1];
   auto layerWeights = layer.weights();
   auto& w = *layerWeights[0];

   w[0] =  -0.1;
   w[1] =   0.2;
   w[2] =  -0.3;
   w[3] =   0.4;

   std::cout << "Trainable: " << trainable << std::endl;
   std::cout << "Weights before adding layer: " << w[0] << " " << w[1] << " " << w[2] << " " << w[3] << std::endl;

   layer.set_trainable(trainable);
   net << tiny_dnn::fc<tiny_dnn::tan_h>(2, 2, false);

   std::cout << "Weights after adding layer: " << w[0] << " " << w[1] << " " << w[2] << " " << w[3] << std::endl;
}

int main() {
   showcase(false);
   showcase(true);
   return 0;
}

```

Output:
```
$ g++ -std=c++11 -pthread -Itiny-dnn tinydnn.cpp 
$ ./a.out
 
Trainable: 0
Weights before adding layer: -0.1 0.2 -0.3 0.4
Weights after adding layer: -0.1 0.2 -0.3 0.4
Trainable: 1
Weights before adding layer: -0.1 0.2 -0.3 0.4
Weights after adding layer: -0.203254 1.21785 0.539683 1.05954
```

**Comment by pliptor** [comment]
Maybe if you have xavier type initialization of weights, and the layer is set to trainable, adding new layers  may readjust of previous weight depending on fan in/fan out changes with the addition of subsequent layers.

**Comment by edgarriba** [comment]
@m-toman I don't have an answer right now but what you could do is try to use a debugger to see what happens internally for each case. /cc @nyanp you implemented `set_trainable` routine, any idea?

**Comment by nyanp** [comment]
@m-toman 
Thank you for clarifying the issue.

#### TL;DR
It is intended behavior. If you want to avoid this behavior, you need to modify weights after constructing all layers and calling ```network::init_weight()``` explicitly.

However I think it's better not to reset weights when adding a new layer.

#### Detailed Description
Each layer has an internal flag (layer::initialized_) to check if weights are initialized. This flag is marked as "initialized" when:

- A layer is imported via caffe-converter
- A layer weights are loaded via ```layer::load``` or ```network::load```
- ```layer::init_weight()``` is explicitly called by user

In other cases, tiny-dnn uses "lazy initialization" strategy on weight initialization. This means that **when we need weights of a layer AND it is not initialized yet, the layer allocates and initializes weights implicitly**. Now this will occur when

1. Adding a new layer
1. Execute forward or backward pass

...but we might be able to avoid weight initialization in 1.

**Comment by m-toman** [comment]
@nyanp 
Thanks for the helpful answer. Would have spared me quite some time if I asked before debugging for days :).
I've checked how it's done in the caffe-converter but seem I missed that part.

So in my case I could do:
- add all layers
- call init_weight
- overwrite the initialized weights by accessing them with layer.weights()

Correct?

**Comment by nyanp** [comment]
@m-toman 
Correct. Contribution to improve this behavior is welcome :)
